### DISCUSSION TITLE
Using PDO ClientPool in the HttpServer?

### DISCUSSION THREAD

**Discussion by vrgenl** [swoole/swoole-src#5122] [General]
I'm wondering what the best practice is for the PDO ClientPool in combination with the HttpServer. I need the $pool->get() in the onRequest and would like to share those connections throughout all HttpWorkers or do misunderstand those pools?

Any help is appreciated.

Kind regards,

Rogier

**Comment by vrgenl** [comment]
Let me elaborate, at what point do i create the pool? I can't do it before creating the server and onrequest creates multiple pools.

**Reply by Draghmar** [reply]
You should create a pool before starting the server. Then you should take connection from the pool when needed and set it back when it's not. So when there's a `onRequest` that requires DB, you should use `$conn = $pool->get()`, do some stuff with it and then use `$pool->put($conn)` before `$response->end()`. An simplified example based on my code:
```php
	$dbpool = new ConnectionPoolExt(function() use ($config) {
		return new DB($config);
	}, DBPOOL);

	$srv = new Server(SWOOLEADDR, SWOOLEPORT);
	$srv->set([
		'log_level' => 0,
		'log_file' => '/var/log/swoole/swoole.log',
		'log_rotation' => SWOOLE_LOG_ROTATION_DAILY,
		'reload_async' => true,
		'http_compression_level' => 5,
		'http_gzip_level' => 5,
		'http_parse_post' => true,
		'task_worker_num' => swoole_cpu_num()*TASKPOOL,
		'task_enable_coroutine' => true,
		'stats_file' => '/var/log/swoole/stats.log',
	]);
	$srv->on('request', function($request, $response) use ($srv, $dbpool) {
		try {
			$db = $dbpool->get();
		} catch(Throwable $e) {
			$dbpool->put($db);
			return $response->end();
		}
		// Do some DB stuff
		$dbpool->put($db);
		$response->end();
	});
	$srv->start();
```
